File: src/browser/git/refs.ts
import * as path from "path";
import * as fs from "fs-extra";
import { exec } from "./exec";

function addRef(refs: Refs, r: Ref): void {
  (refs.refsById[r.id] || (refs.refsById[r.id] = [])).push(r);
  switch (r.type) {
    case "HEAD":
      refs.head = r.id;
      break;
    case "MERGE_HEAD":
      refs.mergeHeads.push(r.id);
      break;
    case "heads":
      refs.heads.push(r);
      break;
    case "remotes":
      (refs.remotes[r.remote] || (refs.remotes[r.remote] = [])).push(r);
      break;
    case "tags":
      refs.tags.push(r);
      break;
    default:
      throw new Error("Unknown ref type");
  }
}

function parseForEachRefLine(line: string): Ref | undefined {
  const tokens = line.split("\0");
  if (tokens.length !== 5) {
    throw new Error("for-each-ref: Unexpected output: " + line);
  }
  const [id, head, dereferenced, objectType, fullname] = tokens;
  const refNameComponents = fullname.split("/");
  let type = refNameComponents.shift();
  if (type === "HEAD") {
    return { fullname, type, id };
  } else if (type === "refs") {
    type = refNameComponents.shift();
    if (type === "heads") {
      const name = refNameComponents.join("/");
      const current = head === "*";
      return { fullname, type, name, id, current };
    } else if (type === "tags") {
      const name = refNameComponents.join("/");
      if (objectType === "tag") {
        return { fullname, type, name, id: dereferenced, tagId: id };
      } else {
        return { fullname, type, name, id, tagId: id };
      }
    } else if (type === "remotes") {
      const remote = refNameComponents.shift();
      if (remote) {
        const name = refNameComponents.join("/");
        return { fullname, type, remote, name, id };
      } else {
        throw new Error("for-each-ref: Unexpected output: " + line);
      }
    } else {
      return undefined;
    }
  } else {
    throw new Error("for-each-ref: Unexpected output: " + line);
  }
}

const SortOrder: { [key in Ref["type"]]: number } = {
  HEAD: 0,
  heads: 1,
  tags: 2,
  remotes: 3,
  MERGE_HEAD: 4
};

export async function getRefs(repository: string): Promise<Refs> {
  const refs: Refs = {
    mergeHeads: [],
    remotes: {},
    heads: [],
    tags: [],
    refsById: {}
  };
  await exec("for-each-ref", {
    repository,
    args: [
      "--sort",
      "-creatordate",
      "--format",
      "%(objectname)%00%(HEAD)%00%(*objectname)%00%(objecttype)%00%(refname)"
    ],
    onEachLine: (line) => {
      const ref = parseForEachRefLine(line);
      if (ref) {
        addRef(refs, ref);
        if (ref.type === "heads" && ref.current) {
          addRef(refs, { type: "HEAD", id: ref.id, fullname: "HEAD" });
        }
      }
    }
  });
  if (!refs.head) {
    // maybe HEAD is detached
    const ret = await exec("rev-parse", {
      repository,
      args: ["HEAD"],
      allowNonZeroExitCode: true
    });
    if (ret.exitCode === 0) {
      const headId = ret.stdout.toString("utf8").replace(/\n$/, "");
      addRef(refs, { type: "HEAD", id: headId, fullname: "HEAD" });
    }
  }
  // Get merge head list from .git/MERGE_HEAD
  const mergeHeadsFile = path.join(repository, ".git/MERGE_HEAD");
  if (await fs.pathExists(mergeHeadsFile)) {
    const type = "MERGE_HEAD";
    const data = await fs.readFile(mergeHeadsFile);
    data
      .toString()
      .replace(/\n$/, "")
      .split("\n")
      .forEach((id, index) => addRef(refs, { fullname: `MERGE_HEAD/${index}`, type, id }));
  }
  // sort each refsById entries
  Object.keys(refs.refsById).forEach((key) => {
    refs.refsById[key].sort((a, b) => SortOrder[a.type] - SortOrder[b.type]);
  });
  return refs;
}

File: src-tauri/app/src/git/refs.rs
use super::{exec, merge_heads, rev_parse, GitError};
use std::path::Path;
use types::*;

fn parse_refs_output(output: &str) -> Result<Refs, GitError> {
    let mut refs = Refs {
        head: None,
        merge_heads: Vec::new(),
        refs: Vec::new(),
    };
    for line in output.lines() {
        let tokens: Vec<&str> = line.split('\0').collect();
        match tokens.as_slice() {
            [sha, head, deref, object_type, fullname] => {
                let refname_components: Vec<&str> = fullname.split('/').collect();
                if refname_components[0] != "refs" {
                    continue;
                }
                match refname_components[1] {
                    "heads" => {
                        let branch = Ref::Branch {
                            fullname: fullname.to_string(),
                            id: sha.to_string(),
                            name: refname_components[2..].join("/").to_string(),
                            current: *head == "*",
                        };
                        refs.refs.push(branch);
                        if *head == "*" {
                            refs.head = Some(sha.to_string());
                        }
                    }
                    "tags" => {
                        refs.refs.push(Ref::Tag {
                            fullname: fullname.to_string(),
                            id: (if *object_type == "tag" { deref } else { sha }).to_string(),
                            tag_sha: sha.to_string(),
                            name: refname_components[2..].join("/").to_string(),
                        });
                    }
                    "remotes" => {
                        let remote = refname_components[2];
                        refs.refs.push(Ref::Remote {
                            fullname: fullname.to_string(),
                            id: sha.to_string(),
                            remote: remote.to_string(),
                            name: refname_components[3..].join("/").to_string(),
                        });
                    }
                    _ => {
                        // do nothing
                    }
                }
            }
            _ => {
                return Err(GitError::UnexpectedOutput {
                    command: String::from("for-each-ref"),
                    text: line.to_string(),
                })
            }
        }
    }
    Ok(refs)
}

pub async fn get_refs(repo_path: &Path) -> Result<Refs, GitError> {
    let repo_path = Path::new(repo_path);
    let args = vec![
        "--sort",
        "-creatordate",
        "--format",
        "%(objectname)%00%(HEAD)%00%(*objectname)%00%(objecttype)%00%(refname)",
    ];
    let output = exec(repo_path, "for-each-ref", &args, &[]).await?;
    GitError::assert_process_output("for-each-ref", &output)?;
    let stdout = std::str::from_utf8(&output.stdout).unwrap();
    let mut refs = parse_refs_output(stdout)?;
    if refs.head.is_none() {
        refs.head = rev_parse::rev_parse(repo_path, "HEAD").await?;
    }
    refs.merge_heads = merge_heads::get_merge_heads(repo_path).await?;
    Ok(refs)
}

pub async fn get_reflog(repo_path: &Path, count: u32) -> Result<Vec<(String, String)>, GitError> {
    if count == 0 {
        return Ok(Vec::new());
    }
    let count_option = format!("-{}", count);
    let args = vec!["--format=%gd %H", "-z", &count_option];
    let output = exec(repo_path, "reflog", &args, &[]).await?;
    GitError::assert_process_output("reflog", &output)?;
    let stdout = std::str::from_utf8(&output.stdout).unwrap();
    let lines = stdout.split('\0').collect::<Vec<_>>();
    let ret = lines
        .iter()
        .filter(|&l| l.len() > 0)
        .map(|&l| {
            let tokens = l.split(' ').collect::<Vec<_>>();
            (tokens[0].to_string(), tokens[1].to_string())
        })
        .collect::<Vec<_>>();
    Ok(ret)
}
